src/Context/UserInfo.tsx

```diff
@@ -0,0 +1,122 @@
+import { createContext, useContext, useReducer } from "react";
+
+// Define Type for UserContextProvider
+type UserProps = {
+  children: React.ReactNode;
+};
+
+//Define Car
+interface UICar {
+  id: number,
+  brand: string, 
+  color: string
+}
+
+// Define Ui Person Interface
+interface UIPerson {
+  name: string;
+  lastName: string;
+  isLoggedIn: boolean;
+  cars: {id: number, brand: string, color: string}[];
+  toggleLogin?: () => void;
+  addUserCar?: (data: UICar) => void;
+  deleteUserCar?: (data: UICar) => void;
+  updateUserCar?: (data: UICar) => void;
+}
+
+// Define Basic Actions Type "Generic"
+enum ActionKind{
+  logged = 'logged',
+  add = 'add',
+  update = 'update',
+  delete = 'delete'
+}
+
+type Action = {
+  type: ActionKind,
+  payload: { id: number, brand: string, color: string}
+}
+
+
+const carReducer = ( state: UIPerson, action: Action ) =>{
+  switch (action.type) {
+    case ActionKind.add:
+      return {
+        ...state,
+        cars: [ ...state.cars , { ...action.payload } ]
+      }
+    case ActionKind.delete:
+      // Require id in the Action.Payload
+      return {
+        ...state,
+        cars: state.cars.filter( car => car.id!== action.payload.id)
+      }
+    case ActionKind.update:
+      // Require id in the Action.Payload
+      const carsFilter = state.cars.filter( car => car.id!== action.payload.id)
+      return{
+        ...state,
+        cars: [ ...carsFilter, action.payload ]
+      }
```

Comment:
this updates the car's array, however, the updated car loses its position in the array if is not the last. you could do this to fix it:
```
const indexToUpdate = state.findIndex((car) => car.id === action.payload.id)
return {
   ... state,
   cars: [ ...state.cars.slice(0, indexToDelete), action.payload, ...state.cars.slice(indexToUpdate + 1)]
}
```